File: R/02_modelos_bayesianos.R
# =============================================================================
# Script: 02_modelos_bayesianos.R
# Descripción: Plantilla para análisis bayesiano de datos experimentales
#              biológicos / agronómicos con brms (Stan backend).
#              Modelos: regresión lineal, ANOVA bayesiano, GLMM bayesiano.
# Autor: [nombre]
# Fecha: [fecha]
# =============================================================================

# ── 0. Paquetes ──────────────────────────────────────────────────────────────
if (!requireNamespace("pacman", quietly = TRUE)) install.packages("pacman")
pacman::p_load(
  tidyverse,   # manipulación y visualización
  brms,        # modelos bayesianos via Stan
  bayesplot,   # gráficos de diagnóstico MCMC
  tidybayes,   # extracción y visualización de distribuciones posteriores
  ggdist,      # distribuciones para ggplot2
  posterior,   # resumen de draws
  loo          # comparación de modelos (LOO-CV)
)

# ── 1. Datos de ejemplo ──────────────────────────────────────────────────────
set.seed(42)
datos <- data.frame(
  tratamiento = rep(c("Control", "Dosis_baja", "Dosis_alta"), each = 20),
  bloque      = factor(rep(1:4, times = 15)),
  rendimiento = c(
    rnorm(20, mean = 5.0, sd = 0.8),
    rnorm(20, mean = 6.2, sd = 0.9),
    rnorm(20, mean = 7.5, sd = 1.0)
  ),
  conteo      = rpois(60, lambda = rep(c(3, 6, 10), each = 20))
)
datos$tratamiento <- factor(datos$tratamiento,
                            levels = c("Control", "Dosis_baja", "Dosis_alta"))

# ── 2. Priors informativos ───────────────────────────────────────────────────
# Explorar priors por defecto para el modelo
get_prior(rendimiento ~ tratamiento + (1 | bloque), data = datos, family = gaussian())

# Definir priors propios (ajustar según conocimiento del dominio)
priors_lmm <- c(
  prior(normal(5, 3),  class = "Intercept"),
  prior(normal(0, 2),  class = "b"),
  prior(exponential(1), class = "sd"),
  prior(exponential(1), class = "sigma")
)

# ── 3. Modelo lineal mixto bayesiano ─────────────────────────────────────────
cat("\n====== LMM Bayesiano ======\n")
modelo_bayes_lmm <- brm(
  formula  = rendimiento ~ tratamiento + (1 | bloque),
  data     = datos,
  family   = gaussian(),
  prior    = priors_lmm,
  chains   = 4,
  iter     = 4000,
  warmup   = 1000,
  cores    = parallel::detectCores() - 1,
  seed     = 42,
  file     = "resultados/modelo_bayes_lmm"  # caché del modelo compilado
)

# Resumen del modelo
summary(modelo_bayes_lmm)
print(modelo_bayes_lmm, digits = 3)

# ── 4. Diagnósticos MCMC ─────────────────────────────────────────────────────
# Trazas y distribuciones
plot(modelo_bayes_lmm)

# R-hat y tamaño efectivo de muestra
bayesplot::mcmc_rhat(brms::rhat(modelo_bayes_lmm))
bayesplot::mcmc_neff(brms::neff_ratio(modelo_bayes_lmm))

# Posterior predictive check
pp_check(modelo_bayes_lmm, ndraws = 100)
pp_check(modelo_bayes_lmm, type = "boxplot", ndraws = 50)

# ── 5. Inferencia bayesiana ───────────────────────────────────────────────────
# Distribuciones posteriores de los coeficientes
posterior_samples <- as_draws_df(modelo_bayes_lmm)

# Intervalos de credibilidad al 95% (HDI)
posterior::summarise_draws(
  posterior_samples,
  mean, median,
  ~posterior::quantile2(.x, probs = c(0.025, 0.975)),
  posterior::default_convergence_measures()
)

# Probabilidad de que el efecto de Dosis_alta > 0
mean(posterior_samples$b_tratamientoDosis_alta > 0)

# ── 6. Comparaciones entre tratamientos ──────────────────────────────────────
# Medias marginales posteriores
em_bayes <- emmeans::emmeans(modelo_bayes_lmm, ~ tratamiento)
summary(em_bayes, point.est = mean)

# Contrastes
emmeans::contrast(em_bayes, method = "pairwise")

# Con tidybayes
datos %>%
  tidybayes::add_epred_draws(modelo_bayes_lmm) %>%
  ggplot(aes(x = .epred, y = tratamiento, fill = tratamiento)) +
  ggdist::stat_halfeye(.width = c(0.66, 0.95)) +
  labs(
    title = "Distribución posterior de medias por tratamiento",
    x     = "Rendimiento esperado",
    y     = "Tratamiento"
  ) +
  theme_bw(base_size = 13) +
  theme(legend.position = "none")

# ── 7. GLMM bayesiano (Poisson) ───────────────────────────────────────────────
cat("\n====== GLMM Bayesiano - Poisson ======\n")
priors_glmm <- c(
  prior(normal(0, 2.5), class = "Intercept"),
  prior(normal(0, 1),   class = "b"),
  prior(exponential(1), class = "sd")
)

modelo_bayes_glmm <- brm(
  formula  = conteo ~ tratamiento + (1 | bloque),
  data     = datos,
  family   = poisson(link = "log"),
  prior    = priors_glmm,
  chains   = 4,
  iter     = 4000,
  warmup   = 1000,
  cores    = parallel::detectCores() - 1,
  seed     = 42,
  file     = "resultados/modelo_bayes_glmm_poisson"
)

summary(modelo_bayes_glmm)
pp_check(modelo_bayes_glmm, ndraws = 100)

# ── 8. Comparación de modelos (LOO-CV) ────────────────────────────────────────
cat("\n====== Comparación de modelos ======\n")

# Modelo nulo (solo intercepto + bloque)
modelo_nulo <- brm(
  formula = rendimiento ~ 1 + (1 | bloque),
  data    = datos,
  family  = gaussian(),
  prior   = c(
    prior(normal(5, 3),   class = "Intercept"),
    prior(exponential(1), class = "sd"),
    prior(exponential(1), class = "sigma")
  ),
  chains = 4, iter = 4000, warmup = 1000,
  cores  = parallel::detectCores() - 1,
  seed   = 42,
  file   = "resultados/modelo_bayes_nulo"
)

# LOO
loo_lmm  <- loo(modelo_bayes_lmm)
loo_nulo <- loo(modelo_nulo)
loo_compare(loo_lmm, loo_nulo)

# ── 9. Exportar resultados ────────────────────────────────────────────────────
# Descomentar para guardar
# saveRDS(modelo_bayes_lmm, "resultados/modelo_bayes_lmm.rds")
# write.csv(
#   as.data.frame(fixef(modelo_bayes_lmm)),
#   "resultados/efectos_fijos_bayes.csv"
# )

File: Python/02_modelos_bayesianos.py
"""
Script: 02_modelos_bayesianos.py
Descripción: Plantilla para análisis bayesiano de datos experimentales
             biológicos / agronómicos con PyMC.
             Modelos: regresión lineal bayesiana, ANOVA bayesiano, GLMM bayesiano.
Autor: [nombre]
Fecha: [fecha]
"""

# ── 0. Importaciones ──────────────────────────────────────────────────────────
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import arviz as az          # pip install arviz
import pymc as pm           # pip install pymc

# Configuración
np.random.seed(42)
az.style.use("arviz-darkgrid")

# ── 1. Datos de ejemplo ───────────────────────────────────────────────────────
n_per_group = 20
n_bloques = 4
tratamientos_str = np.repeat(["Control", "Dosis_baja", "Dosis_alta"], n_per_group)
bloques_num = np.tile(np.arange(n_bloques), 15)

rendimiento = np.concatenate([
    np.random.normal(5.0, 0.8, n_per_group),
    np.random.normal(6.2, 0.9, n_per_group),
    np.random.normal(7.5, 1.0, n_per_group),
])
conteo = np.random.poisson(lam=np.repeat([3, 6, 10], n_per_group))

# Codificación numérica de tratamiento
trat_map = {"Control": 0, "Dosis_baja": 1, "Dosis_alta": 2}
tratamiento_idx = np.array([trat_map[t] for t in tratamientos_str])
n_tratamientos = len(trat_map)

# ── 2. Modelo lineal bayesiano (ANOVA bayesiano) ─────────────────────────────
print("\n====== Modelo Lineal Bayesiano (ANOVA) ======")

with pm.Model(coords={"tratamiento": list(trat_map.keys())}) as modelo_bayes_lm:

    # Priors
    mu_tratamiento = pm.Normal("mu_tratamiento", mu=5.0, sigma=3.0,
                               dims="tratamiento")
    sigma = pm.HalfNormal("sigma", sigma=2.0)

    # Esperanza
    mu_obs = mu_tratamiento[tratamiento_idx]

    # Verosimilitud
    y_obs = pm.Normal("y_obs", mu=mu_obs, sigma=sigma, observed=rendimiento)

    # Muestreo
    traza_lm = pm.sample(
        draws=2000,
        tune=1000,
        chains=4,
        cores=4,
        random_seed=42,
        target_accept=0.9,
        return_inferencedata=True
    )

# ── 3. Diagnósticos MCMC ─────────────────────────────────────────────────────
print("\nDiagnósticos MCMC:")
print(az.summary(traza_lm, var_names=["mu_tratamiento", "sigma"]))

# Trazas
az.plot_trace(traza_lm, var_names=["mu_tratamiento", "sigma"])
plt.tight_layout()
plt.savefig("resultados/trazas_modelo_lm_bayes.png", dpi=150)
plt.show()

# Distribuciones posteriores
az.plot_posterior(traza_lm, var_names=["mu_tratamiento"])
plt.savefig("resultados/posterior_mu_tratamientos.png", dpi=150)
plt.show()

# Posterior predictive check
with modelo_bayes_lm:
    ppc = pm.sample_posterior_predictive(traza_lm, random_seed=42)

az.plot_ppc(az.from_pymc(
    posterior_predictive=ppc,
    model=modelo_bayes_lm
), num_pp_samples=100)
plt.savefig("resultados/ppc_modelo_lm_bayes.png", dpi=150)
plt.show()

# ── 4. Contrastes entre tratamientos ─────────────────────────────────────────
post_mu = traza_lm.posterior["mu_tratamiento"]

diferencia_alta_control = (
    post_mu.sel(tratamiento="Dosis_alta") -
    post_mu.sel(tratamiento="Control")
)
prob_pos = float((diferencia_alta_control > 0).mean())
print(f"\nP(Dosis_alta > Control) = {prob_pos:.3f}")

# HDI al 94%
hdi_diff = az.hdi(diferencia_alta_control.values, hdi_prob=0.94)
print(f"HDI 94% de la diferencia: [{hdi_diff[0]:.3f}, {hdi_diff[1]:.3f}]")

# ── 5. Modelo Lineal Mixto Bayesiano (LMM) ───────────────────────────────────
print("\n====== LMM Bayesiano (efecto aleatorio de bloque) ======")

with pm.Model(coords={
    "tratamiento": list(trat_map.keys()),
    "bloque": [str(i) for i in range(n_bloques)]
}) as modelo_lmm_bayes:

    # Priors para efectos fijos
    alpha = pm.Normal("alpha", mu=5.0, sigma=3.0)
    beta  = pm.Normal("beta", mu=0.0, sigma=2.0, dims="tratamiento")

    # Prior para efectos aleatorios de bloque
    sigma_bloque = pm.HalfNormal("sigma_bloque", sigma=1.0)
    u_bloque = pm.Normal("u_bloque", mu=0.0, sigma=sigma_bloque, dims="bloque")

    sigma_resid = pm.HalfNormal("sigma_resid", sigma=2.0)

    mu_obs = alpha + beta[tratamiento_idx] + u_bloque[bloques_num]

    y_obs = pm.Normal("y_obs", mu=mu_obs, sigma=sigma_resid, observed=rendimiento)

    traza_lmm = pm.sample(
        draws=2000,
        tune=1000,
        chains=4,
        cores=4,
        random_seed=42,
        target_accept=0.9,
        return_inferencedata=True
    )

print(az.summary(traza_lmm, var_names=["alpha", "beta", "sigma_bloque", "sigma_resid"]))

az.plot_forest(traza_lmm, var_names=["beta"], combined=True,
               hdi_prob=0.95, r_hat=True)
plt.title("Efectos de tratamiento – LMM Bayesiano")
plt.savefig("resultados/forest_beta_lmm_bayes.png", dpi=150)
plt.show()

# ── 6. GLMM Bayesiano – Poisson ───────────────────────────────────────────────
print("\n====== GLMM Bayesiano - Poisson ======")

with pm.Model(coords={
    "tratamiento": list(trat_map.keys()),
    "bloque": [str(i) for i in range(n_bloques)]
}) as modelo_glmm_pois:

    alpha_p = pm.Normal("alpha_p", mu=1.0, sigma=1.5)
    beta_p  = pm.Normal("beta_p",  mu=0.0, sigma=1.0, dims="tratamiento")

    sigma_b = pm.HalfNormal("sigma_b", sigma=0.5)
    u_b = pm.Normal("u_b", mu=0.0, sigma=sigma_b, dims="bloque")

    log_mu = alpha_p + beta_p[tratamiento_idx] + u_b[bloques_num]
    mu_p = pm.math.exp(log_mu)

    y_cnt = pm.Poisson("y_cnt", mu=mu_p, observed=conteo)

    traza_glmm_pois = pm.sample(
        draws=2000,
        tune=1000,
        chains=4,
        cores=4,
        random_seed=42,
        target_accept=0.9,
        return_inferencedata=True
    )

print(az.summary(traza_glmm_pois, var_names=["alpha_p", "beta_p", "sigma_b"]))

# ── 7. Comparación de modelos (LOO-CV) ────────────────────────────────────────
print("\n====== Comparación de modelos (LOO) ======")
loo_lm  = az.loo(traza_lm,  pointwise=True)
loo_lmm = az.loo(traza_lmm, pointwise=True)
comp = az.compare({"LM_bayes": traza_lm, "LMM_bayes": traza_lmm})
print(comp)

az.plot_compare(comp)
plt.savefig("resultados/comparacion_modelos_loo.png", dpi=150)
plt.show()

# ── 8. Exportar resultados ────────────────────────────────────────────────────
# Guardar traza como NetCDF (formato estándar ArviZ)
# traza_lmm.to_netcdf("resultados/traza_lmm_bayes.nc")

# Exportar resumen a CSV
# summary_df = az.summary(traza_lmm).reset_index()
# summary_df.to_csv("resultados/resumen_lmm_bayes.csv", index=False)
